### PR TITLE
search: use exclude filters when resolving global private repos

### LINKED ISSUE #27044: Exclude repo filter doesn't exclude private repos
- **Sourcegraph version:** 3.33.0 
- **Platform information:** Kubernetes, Docker Compose

#### Steps to reproduce:

1. Run a query with the `-repo:` filter to exclude a private repo

#### Expected behavior:
The private repo is excluded from the results.

#### Actual behavior:
The private repo is included in the results. 
__

The `-repo:` filter seems to exclude public repos as intended on 3.33.0.

### PR DESCRIPTION
**Pull request by keegancsmith**
Zoekt is not aware of user permissions, so we calculate the set of
private repos a user can see and pass that on to zoekt. This is only
done in "global mode", otherwise we just use the repo resolver. We
recently added support to exclude repos in global searches. However,
that was only done for the public repositories and we didn't update the
logic which calculated the set of private repositories to take that into
account. This commit adds in the ExcludePattern similiar to how the
repository resolver works.

Note: we should just use the repository resolver. However, we use
slightly different logic for setting "UserID" in the search than the
repo resolver. This commit does the simplest fix, instead of the more
general fix.

Fixes #27044

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_results.go [hunk 1]
```diff
@@ -1597,14 +1598,19 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
...
-			OnlyArchived: args.RepoOptions.OnlyArchived,
-			NoArchived:   args.RepoOptions.NoArchived,
+			UserID:         userID, // Zero valued when not in sourcegraph.com mode
+			OnlyPrivate:    true,
+			LimitOffset:    &database.LimitOffset{Limit: search.SearchLimits(conf.Get()).MaxRepos + 1},
+			OnlyForks:      args.RepoOptions.OnlyForks,
+			NoForks:        args.RepoOptions.NoForks,
+			OnlyArchived:   args.RepoOptions.OnlyArchived,
+			NoArchived:     args.RepoOptions.NoArchived,
+			ExcludePattern: repos.UnionRegExps(args.RepoOptions.MinusRepoFilters),
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_results.go:1613]:
```suggestion
			ExcludePattern: args.RepoOptions.ExcludePattern,
```

**Review by tsenart — APPROVED** [review]

**Review by rvantonder — APPROVED** [review]
> Note: we should just use the repository resolver. However, we use
slightly different logic for setting "UserID" in the search than the
repo resolver.

It would be fantastic if we just used the repo resolver/cursor for this, because then converting global searches to the generic search jobs won't affect the interface. Currently it will, once I migrate it, because the type of the data here is different from repo resolver/cursor. Just flagging--I will probably go ahead with changes regardless and hopefully we can unify the datatype here at some point.

**Review by rvantonder — COMMENTED** [review]
Also, seems like a key thing to add a regression for in integration tests.

**Comment by keegancsmith** [comment]
> > Note: we should just use the repository resolver. However, we use
> > slightly different logic for setting "UserID" in the search than the
> > repo resolver.
> 
> It would be fantastic if we just used the repo resolver/cursor for this, because then converting global searches to the generic search jobs won't affect the interface. Currently it will, once I migrate it, because the type of the data here is different from repo resolver/cursor. Just flagging--I will probably go ahead with changes regardless and hopefully we can unify the datatype here at some point.

I agree, this is just a good thing to do. But I wasn't prepared to onboard all the context and risk more regressions. So I did the smallest fix possible :)

However, this all sounds like it should just work. Note: global searches don't rely on the DB to get the list of repos, only to find out the private repos for the user. So if there is a way to limit the subset of RepoOptions this sounds doable.

I also question why UserID stuff is different. But I lack context here, I think there was some decision made around only including private repos explicitly added by a user va any private repo they have access too. This is why I didn't dive into this rabbithole.

> Also, seems like a key thing to add a regression for in integration tests.

I'll do this.

File: cmd/frontend/graphqlbackend/search_results.go (hunk 1 above)
Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/search_results.go]: This option doesn't exist. We need to do it this way.

**Pull request merged**